BuildCheck polishing one-pager (#11036)
dotnet/msbuild: documentation/specs/proposed/BuildCheck-net10-polishing.md

# BuildCheck Net10 polishing 

In net 9.0 we delivered initial preview of BuildCheck feature. 
 * Spec: https://github.com/dotnet/msbuild/blob/main/documentation/specs/BuildCheck/BuildCheck.md
 * Work track:  https://github.com/orgs/dotnet/projects/373

> In addition to that we have item tracking possible improvemnts and extensions of the feature: https://github.com/dotnet/msbuild/issues/10548

Reviewer: ```suggestion
In addition to that we have an item tracking possible improvements and extensions of the feature: https://github.com/dotnet/msbuild/issues/10548
```

> This doc focuses on minimal investment path that would still allow driving some adoption of BuildChecks feature and collecting feedback from real life usage.

Reviewer: ```suggestion
This doc focuses on a minimal investment path that would allow driving adoption of the BuildChecks feature and collecting feedback from real life usage.
```

## Goals and Motivation 

* Making sure the feature can be used in the real life scenarios.
* Driving adoption.

## Impact 

* MSBuild team has a venue to evangelise best practices for the build.
* Customers have a tool to formalize and distribute their view of best practices.

> * MSBuild team can be improving builds perf and security via providing new Checks notyfying about problems.

Reviewer: ```suggestion
* MSBuild team can improving end-user build perf and security with high leverage by providing new Checks notifying about problems.
```

## Stakeholders 
- PM (@baronfel) - as a customer advocate
- Selected internal partner repo owners (details https://github.com/dotnet/msbuild/issues/10726)

### Successful handover

> - Internal partner teams ideas around BuildChecks collected and considered.

Reviewer: ```suggestion
- Internal partner teams' ideas around BuildChecks collected and considered.
```

> - Internal partner teams' ideas around BuildChecks collected and considered.

Reviewer 2: Do we have any idea on how many checks are we aiming to implement / have capacity for?

Author: Once there is a nonzero allocation of forces - we can start considering :-)

- Selected internal partner teams are using BuildChecks in their build pipelines.
- BuildChecks being evangelized externaly, adoption numbers grow up.

## Risks 
- Performance degradation is unacceptable on real scale project.

> - There are multiple hard to investigate or/and hard to fix bugs identified during inital adoptions.

Reviewer: ```suggestion
- There are multiple hard to investigate and/or hard to fix bugs identified during initial adoptions.
```

> - Unsatisfactory specificity or/and sensitivity - leading to diminishing the valua added and endangering the adoption motivation.

Reviewer: ```suggestion
- Unsatisfactory specificity and/or sensitivity - leading to diminishing the perceived value of Checks and endangering the motivation to adopt them.
```

- Low perceived value of best practices enforced by the checks.
- Confusing diagnostics/documentation leading to lowering the educational value.

## Scope

### Goals
* Running buildchecks on partner repos, identifying and fixing issues

> * Enabling buildchecks on partner repos, providing initial support

Reviewer: Is the distinction here "we try it for them" and then "they turn it on in prod"?

Author: I updated the wording - and atempted to make it more explicit and less cryptic :-)

> * MSBuild team runs buildchecks on selected partner repos (via private runs), identifying and fixing issues

Reviewer 2: For the private runs, do you think we should have some automation to the testing or just running it manually on our machines?

Author: Manual for initial determination of possible issues or perf degrafdations.
But the idea is that this leads to buildchecks then being turned on in the internal automation runs of such builds (e.g. in our own CI)

### Non-Goals

> * Measuring and defining the perf impact, detecting the sources of it

Reviewer: Since perf was called out as an explicit risk, why is this a non-goal?

Author: Yes - perf is super important. If there is an appetite to invest any efforts into buildchecks - it should be on top of the list. If there is near to zero investment - then we need to at least keep it alive by dogfooding. Dogfooding can then fuel and justify investment into the perf.

Reviewer 2: PerfStar also has a plan to implement measures for the BuildCheck feature, so we will be keeping an eye on it at least.

Author: Thanks for suggestion!
Added

> * Providing additional helpful low-hanging-fruit checks

Reviewer: Are we saying "we won't do this"? Or "if we do new low-hanging checks it's not really under this umbrella"?

Author: Rather - "We may be adding some, but we are not aiming nor planning for it". Again - result of dogfooding should be informing this.

### Out of scope

* OM/API enriching
* Configuration and other features improvements
* VS, VS-Code integrations

## Cost 

The below plan is expected with 0.25 Dev / Month investment (except for Month #2, that will realistically need ~0.5-2 Dev / Month)

## Plan 
* Month #1 - Running build checks on selected partner repos and identifying issues
* Month #2 - Resolve identified adoption blockers
* Month #2 optional - run perf tests and define perf profile for build with Checks.
* Month #3 - Enabling buildchecks on partner repos, providing initial support

> * Month #4 - Evangelization and driving the adoption in external community

Reviewer: IMO this is too detailed for a "one-pager". But the plan looks pretty good.
